playground/eShopLite/CatalogDb/Program.cs

```diff
@@ -1,9 +1,19 @@
 using CatalogDb;
+using Microsoft.EntityFrameworkCore;
+using Npgsql;
 
 var builder = WebApplication.CreateBuilder(args);
 
 builder.AddServiceDefaults();
-builder.AddNpgsqlDbContext<CatalogDbContext>("catalogdb");
+
+var connectionString = builder.Configuration.GetConnectionString("catalogdb")
+    ?? throw new InvalidOperationException("Connection string is not configured.");
+
+// a workaround for https://github.com/npgsql/efcore.pg/issues/2821
+var configureNpgsqlLogging = (NpgsqlDataSourceBuilder builder) => { builder.UseLoggerFactory(null); };
+builder.Services.AddNpgsqlDataSource(connectionString, configureNpgsqlLogging)
```

Comment:
Is the `AddNpgsqlDataSource` call needed because `DbDataSource`/`DbConnection` is used somewhere without going through EF?

Reply:
IMHO this is generally good setup practice even if ADO.NET isn't used directly - 

However, there are unfortunately some issues with having the data source pulled directly out of the application service provider (by calling UseNpgsql() below without an argument) - see https://github.com/npgsql/efcore.pg/issues/2891. This is something I definitely want to fix for 9.0, but it's not easy.

In the meantime, instead of using AddNpgsqlDataSource it may be better to directly instantiate an NpgsqlDataSource and then pass it to UseNpgsql(); note that this would also obviate the logging workaround above as well, since the data source would no longer automatically pick up the logging factory from DI. So something like the following:

```c#
var npgsqlDataSource = NpgsqlDataSource.Create(connectionString);
```

Note that NpgsqlDataSource is disposable, but disposing it here causes issues with the way migrations work. In any case, since there's just one instance for the entire lifetime of the application, it should be fine.